Move to YAML for stylesheets

lornajane opened the pull request:
Fixes #876 where we've been discussing moving away from RSON to something a bit more widely used/supported/familiar for users.

This PR introduces:
 - Support for stylesheets with `.yaml` and `.yml` extensions, in valid YAML format
 - A converter tool in `rst2pdf/style-to-yaml.py` to convert any existing stylesheet (including our default ones)

Other things to include before merging if we decide this is a useful thing and we should pursue it:
 - stylesheet documentation needs a big overhaul, also see #906 where the manual examples are in JSON at the moment
 - are there other ways to represent some of those nested table styles? They are UGLY
 - what else?

Please kick the tyres, convert your stylesheets, see what you think and let's decide if this is a good approach.

akrabat commented:
With the merging of #951, you'll need to rebase against `master` in order to pass the CI checks.

lornajane commented:
Rebased as requested

stephenfin commented on rst2pdf/styles.py:
```diff
@@ -610,7 +611,11 @@ def readStyle(self, ssname):
             try:
-                return rson_loads(open(fname).read())
+                # Is it an older rson/json stylesheet with .style extension?
+                if fname[-5:] == "style":
```
Could we use `os.path.splitext` to get the extension instead of this? This was trigger for e.g. `foo.mystyle` too.

lornajane replied:
I couldn't figure out the "Python" way for this, thanks for the tip!

stephenfin commented on rst2pdf/styles.py:
```diff
@@ -610,7 +611,11 @@ def readStyle(self, ssname):
-                return rson_loads(open(fname).read())
+                # Is it an older rson/json stylesheet with .style extension?
+                if fname[-5:] == "style":
+                    return rson_loads(open(fname).read())
```
Could we log a warning here, indicating that users should convert their style files to YAML file?

lornajane replied:
Great idea, added!

stephenfin reviewed (changes requested):
Some comments inine. Could we get some docs too?

lornajane commented:
Braindump in case anyone else has batches of styles to convert:

```
for i in *.style; do base=$(basename $i .style); python3 /path/to/style-to-yaml.py $i > $base.yaml; done
```

stephenfin commented:
> Braindump in case anyone else has batches of styles to convert:
> 
> ```
> for i in *.style; do base=$(basename $i .style); python3 /path/to/style-to-yaml.py $i > $base.yaml; done
> ```

Fwiw, if you renamed the file to use underscores or something else (since dashes aren't valid in a module name), you could do:

```
python -m rst2pdf.style2yaml $i
```

lornajane commented:
@stephenfin this is useful to know - @akrabat already asked me if we could add the converter as an option to the rst2pdf command itself and this seems relevant for when I get to that point too!

lornajane commented:
A few tests failed, but they pass locally for me (python 3.8). Any idea if we can get the output or figure out how to reproduce the failures so I can take another look?

stephenfin commented:
Hmm, looks like the python3.7 test failed and the others were cancelled because of this. They likely failed too. Is there any chance you forgot to add some files or something? If not, maybe split this into multiple PRs. One to add the converter, one to add docs, one to convert the test files and a final one to convert the in-tree styles?

lornajane commented:
This is pretty much working now but I think we should add docs before we merge. I am on the case :)

stephenfin commented on rst2pdf/style-to-yaml.py:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python
```
`python3`

On all Linux distros except for Arch, `python` historically points to `python2.7`. This is changing but it's still safer to be explicit IMO

stephenfin replied:
Also, license information? Even a SPDX header would do the trick.

```
# SPDX-License-Identifier: MIT
```

lornajane replied:
Yes to the license headers! We must try to unblock your PR for those across the rest of the project too

stephenfin commented on rst2pdf/style-to-yaml.py:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python
+'''
```
style nit: could we use double quotes here? That's more typical for multi-line docstrings.

lornajane replied:
Fixed, but half the project uses single ones - decided it made no sense to fix other files in this PR. One for the automated stylechecker tools and a rainy day, perhaps

stephenfin commented on rst2pdf/style-to-yaml.py:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python
+'''
+    Convert older RSON stylesheets to YAML format
```
nit: could you dedent this? No reason for it to be indented.

stephenfin commented on rst2pdf/style-to-yaml.py:
```diff
@@ -0,0 +1,25 @@
+from rson import loads as rloads
+
+if __name__ == '__main__':
+    fname = sys.argv[1]
```
Rather than parsing `argv` directly like this, could we use `argparse`? It's trivial and would allow people to get information on expected input via e.g. `-h`. Also, it would make it easier to convert multiple files at once.

```python3
#!/usr/bin/env python3

"""Convert older RSON stylesheets to YAML format.

Convert the file at PATH from RSON to YAML format.
"""

import argparse
import json
import sys

import yaml

from rst2pdf.dumpstyle import fixstyle

def main():
    parser = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument(
        'path',
        dest='paths',
        metavar='PATH',
        nargs='+',
        help='An RSON-formatted file to convert.',
    )
    args = parser.parse_args()
    for path in args.paths:
        pass # stuff here...
```

lornajane replied:
oh this is so much nicer now we've turned it more into a CLI thing, and I learned something too!

stephenfin commented on rst2pdf/style-to-yaml.py:
```diff
@@ -0,0 +1,25 @@
+if __name__ == '__main__':
+    fname = sys.argv[1]
+    # read rson from a file
+    sstr = open(fname, 'rb').read()
```
You're leaving this file open. You rarely want to use `open` like this. Use a context manager instead:

```python3
with open(fname, 'rb') as fh:
    style_data = fixstyle(rloads(fh.read())
    # ...
```

lornajane replied:
Ah ha! Thanks

stephenfin commented on rst2pdf/style-to-yaml.py:
```diff
@@ -0,0 +1,25 @@
+    reparsed_style = json.loads(json_style)
+
+    yaml_style = yaml.dump(reparsed_style)
+    print(yaml_style)
```
It would be nice to be able to direct this to a file rather than having to do so via the terminal. Perhaps we could have a `--save` flag to dump to `'.'.join((os.path.splitext(fname)[0], 'yaml'))`. If you do this, don't forget to check that the file doesn't exist first and probably error out if so.

lornajane replied:
I think this is needed especially now I have support for multiple files so easily on the input. It was too much for today but I'll take another look when I can

lornajane replied:
I _think_ I've added that now but feedback always welcome. I mean, it does work! But might not be excellent.

stephenfin commented on rst2pdf/styles.py:
```diff
@@ -6,6 +6,7 @@
 import os.path
 import sys
 import re
+import yaml
```
This is a third-party import and should be grouped with the below.

```
stdlib

third-party

first-party
```

I don't think we've anything enforcing this here but that's what PEP8 suggests iirc.

lornajane replied:
ah, thanks

stephenfin commented on rst2pdf/styles.py:
```diff
@@ -610,7 +611,16 @@ def readStyle(self, ssname):
+                # Is it an older rson/json stylesheet with .style extension?
+                root_ext = os.path.splitext(fname)
+                if root_ext[1] == ".style":
+                    log.warn(
```
`log.warn` is deprecated. Use `log.warning`. https://stackoverflow.com/a/15655674

lornajane replied:
Oops! Good catch.

stephenfin reviewed (changes requested):
Loads of suggestions for how to make the conversion tool potentially more useful. I do realise this is a single use tool though so you can ignore most of them if you like. The only hard blocker I have is on the use of `log.warn`, which is deprecated and should not be used. If you can change that at least and add the docs you've suggested then I think I'll be happy.

lornajane commented:
All these comments on the conversion script are helpful! I copied one of our existing files and then made it do what I wanted, I guess I didn't think hard about changing anything that had arrived that worked! A bunch of good advice and I'll have another look

lornajane commented:
OK I think this is in pretty good shape, can we review again and decide if we can move forward? Main points:
 * Doesn't break anything that exists (that I know about)
 * Conversion utility is ready to use
 * Manual is updated to use new syntax, and has information about migrating stylesheets
 * Tests are also updated
 
There is absolutely more that needs doing on the manual but I tried to mostly move/improve what was there without getting too hung up about the overall state of the manual. As long as I haven't made anything worse then I hope we can proceed with this stylesheets change and keep chipping away at the manual (but I've moved everything so we should probably merge this before anyone does anything else radical with docs)

stephenfin commented on rst2pdf/style2yaml.py:
```diff
@@ -0,0 +1,66 @@
+                with open(new_path, 'w') as file:
+                    file.write(yaml_style)
+        else:
+            print(yaml_style)
```
Since we can potentially print multiple documents here, can you print a separator before this line so one can identify the different docs? E.g.

```
---
doc1
---
doc2
```

You could also print a comment stating the original filename just after the separator.

More at https://yaml.org/spec/1.2/spec.html#id2760395

stephenfin commented on rst2pdf/styles.py:
```diff
@@ -610,7 +611,16 @@ def readStyle(self, ssname):
         if fname:
             try:
-                return rson_loads(open(fname).read())
+                # Is it an older rson/json stylesheet with .style extension?
```
A TODO to remove this when we drop RSON support in the future would be helpful.

stephenfin reviewed (approved):
I'm happy with this now. Have two additional nits, but nothing that can't be fixed in a follow-up.

lornajane commented:
OK, minor points addressed, if either of you are happy @stephenfin @akrabat I think we can merge.

akrabat commented:
Testing with my presentations, the converted stylesheets generated the correct PDFs, which is nice!

Formatting YAML is clearly hard as none of the options for the `Dumper`'s `default_flow_style` create what I would ideally want which would be lists in `[]` style and dictionaries on separate lines. As that doesn't seem possible, the `default_flow_style` of `None` seems to be the best choice.

stephenfin commented:
I guess we should look at releasing a new version now. Does 1.0 make sense yet?

akrabat commented:
For 1.0, off the top of my head, all I can think of is the error codes one and maybe assessing the way padding/margins work at the end of a page.

Practically, I think that second one isn't something to wait for though.

lornajane commented:
We said we wanted to update the default stylesheet too, and I'd like us to do that. I'm not sure what else we need to do for a 1.0, but let's start getting our ducks in a row and meeting as many existing PRs as we can include? If the stylesheet update makes it then great. If not, probably not a disaster
